pandas/tseries/tools.py

```diff
@@ -409,22 +429,47 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
             except (ValueError, TypeError):
                 raise e
 
+    # perform the conversion
+    if origin == 'julian':
+        if unit != 'D':
+            raise ValueError("unit must be 'D' for origin='julian'")
+        try:
+            arg = arg - tslib.Timestamp(0).to_julian_date()
+        except:
+            raise ValueError("incompatible 'arg' type for given "
+                             "'origin'='julian'")
     if arg is None:
-        return arg
+        result = arg
     elif isinstance(arg, tslib.Timestamp):
-        return arg
+        result = arg
     elif isinstance(arg, ABCSeries):
         from pandas import Series
         values = _convert_listlike(arg._values, False, format)
-        return Series(values, index=arg.index, name=arg.name)
+        result = Series(values, index=arg.index, name=arg.name)
     elif isinstance(arg, (ABCDataFrame, MutableMapping)):
-        return _assemble_from_unit_mappings(arg, errors=errors)
+        result = _assemble_from_unit_mappings(arg, errors=errors)
     elif isinstance(arg, ABCIndexClass):
-        return _convert_listlike(arg, box, format, name=arg.name)
+        result = _convert_listlike(arg, box, format, name=arg.name)
     elif is_list_like(arg):
-        return _convert_listlike(arg, box, format)
+        result = _convert_listlike(arg, box, format)
+    else:
+        result = _convert_listlike(np.array([arg]), box, format)[0]
 
-    return _convert_listlike(np.array([arg]), box, format)[0]
+    # handle origin
+    if origin not in ['unix', 'julian']:
+        try:
+            offset = tslib.Timestamp(origin) - tslib.Timestamp(0)
+        except tslib.OutOfBoundsDatetime:
+            raise ValueError(
+                "origin {} is Out of Bounds".format(origin))
+        except ValueError:
+            raise ValueError("origin {} cannot be converted "
+                             "to a Timestamp".format(origin))
+
+        if offset is not None:
+            result = result + offset
```

Comment:
As I commented in the original PR (https://github.com/pandas-dev/pandas/pull/11470#issuecomment-251619873), I think it can solve some corner cases to do the origin handling *before* the actual parsing (as you do for the 'julian' case).

A bit an artificial example, but to show one of the problems:

```
In [11]: pd.to_datetime(200*365, unit='D')
Out[11]: Timestamp('2169-11-13 00:00:00')

In [12]: pd.to_datetime(200*365, unit='D', origin='1870-01-01')
Out[12]: Timestamp('2069-11-13 00:00:00')

In [13]: pd.to_datetime(300*365, unit='D', origin='1870-01-01')
...
OutOfBoundsDatetime: cannot convert input with unit 'D'
```

So the last one is actually not an OutOfBounds datetime, but just because we first parse the number as epoch, it raises this error. If we first subtract the correct value from the numeric argument, and then parse as epoch, the above will work fine.